## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :x: [more...](#5dadaace-9154-4a08-8a78-c888e4bd472a) |
| mint-zoned.sh | :x: [more...](#12e8b737-ea71-4772-8085-86abdc2eeeb6) |
| mint-compress-encrypt-dist-erasure.sh | :x: [more...](#19d1b47b-579a-4a98-b484-30b46ed8d9c6) |
<a name="12e8b737-ea71-4772-8085-86abdc2eeeb6"></a>
### 12144-c88cd97/mint-zoned.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev8.minio.io:31935
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 085283c49c50:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 2 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 43 seconds
(4/15) Running aws-sdk-ruby tests ... done in 4 seconds
(5/15) Running awscli tests ... done in 1 minutes and 25 seconds
(6/15) Running healthcheck tests ... done in 1 seconds
(7/15) Running mc tests ... done in 42 seconds
(8/15) Running minio-dotnet tests ... done in 51 seconds
(9/15) Running minio-go tests ... done in 2 minutes and 1 seconds
(10/15) Running minio-java tests ... done in 1 minutes and 59 seconds
(11/15) Running minio-js tests ... done in 58 seconds
(12/15) Running minio-py tests ... FAILED in 2 minutes and 38 seconds
{
  "name": "minio-py:test_thread_safe",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-564d2b16-fe14-4c26-b8a0-1a3eafe74d86",
    "object_name": "eab733a4-67a5-41f1-9cc9-d366fd1db7aa"
  },
  "message": "S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-564d2b16-fe14-4c26-b8a0-1a3eafe74d86, request_id: 1678DB1796F09A64, host_id: be5d06d4-72f0-4e2d-af6a-b41e568a0075, bucket_name: minio-py-test-564d2b16-fe14-4c26-b8a0-1a3eafe74d86",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 126, in _call_test\n    func(log_entry, *args, **kwargs)\n  File \"/mint/run/core/minio-py/tests.py\", line 1640, in test_thread_safe\n    _CLIENT.remove_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 676, in remove_bucket\n    self._execute(\"DELETE\", bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 409, in _execute\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 385, in _url_open\n    raise response_error\nminio.error.S3Error: S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-564d2b16-fe14-4c26-b8a0-1a3eafe74d86, request_id: 1678DB1796F09A64, host_id: be5d06d4-72f0-4e2d-af6a-b41e568a0075, bucket_name: minio-py-test-564d2b16-fe14-4c26-b8a0-1a3eafe74d86\n",
  "duration": 7472
}
(12/15) Running s3cmd tests ... done in 22 seconds
(13/15) Running s3select tests ... done in 5 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
<a name="5dadaace-9154-4a08-8a78-c888e4bd472a"></a>
### 12144-c88cd97/mint-gateway-s3.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev6.minio.io:31945
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 0c2964af8ccf:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 41 seconds
(4/15) Running aws-sdk-ruby tests ... done in 2 seconds
(5/15) Running awscli tests ... done in 2 minutes and 9 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 36 seconds
(8/15) Running minio-dotnet tests ... done in 33 seconds
(9/15) Running minio-go tests ... done in 43 seconds
(10/15) Running minio-java tests ... done in 27 seconds
(11/15) Running minio-js tests ... FAILED in 8 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 11,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:79:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (internal/streams/readable.js:684:10) at endReadableNT (internal/streams/readable.js:1327:12) at processTicksAndRejections (internal/process/task_queues.js:80:21)"
}
(11/15) Running minio-py tests ... done in 1 minutes and 23 seconds
(12/15) Running s3cmd tests ... done in 58 seconds
(13/15) Running s3select tests ... done in 3 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
<a name="19d1b47b-579a-4a98-b484-30b46ed8d9c6"></a>
### 12144-c88cd97/mint-compress-encrypt-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c2.minio.io:31001
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 1b3c21ba98b0:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 43 seconds
(4/15) Running aws-sdk-ruby tests ... done in 3 seconds
(5/15) Running awscli tests ... FAILED in 32 seconds
{
  "name": "awscli",
  "duration": 2584,
  "function": "aws --endpoint-url http://minio-c2.minio.io:31001 s3api copy-object --bucket awscli-mint-test-bucket-12752 --key datafile-1-kB-copy --copy-source awscli-mint-test-bucket-12752/datafile-1-kB\n",
  "status": "FAIL",
  "error": "Hash mismatch expected 084e1383b70fb0c51acc680fef370023, got ac57de7156d7fc25ac1a65f81fa3989b"
}
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... done in 50 seconds
(7/15) Running minio-dotnet tests ... done in 44 seconds
(8/15) Running minio-go tests ... FAILED in 2 minutes and 0 seconds
{
  "args": {},
  "duration": 315,
  "error": "At least one of the pre-conditions you specified did not hold",
  "function": "CopyObjectPart(destination, source)",
  "message": "CopyObjectPart call failed",
  "name": "minio-go: testUnencryptedToSSES3CopyObjectPart",
  "status": "FAIL"
}
(8/15) Running minio-java tests ... FAILED in 2 minutes and 6 seconds
{
  "name": "minio-java",
  "function": "copyObject()",
  "args": "[match etag]",
  "duration": 129,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = PreconditionFailed, message = At least one of the pre-conditions you specified did not hold, bucketName = minio-java-test-3hdu59j, objectName = minio-java-test-624kjr-copy, resource = /minio-java-test-3hdu59j/minio-java-test-624kjr-copy, requestId = 1678DADA15762A4D, hostId = e8604228-7869-4776-93c0-84ea27796188)\nrequest={method=PUT, url=http://minio-c2.minio.io:31001/minio-java-test-3hdu59j/minio-java-test-624kjr-copy, headers=x-amz-copy-source-if-match: 71cff0a060f852067e443ad1e24ae26c-1\nx-amz-copy-source: /minio-java-test-3v9p6sj/minio-java-test-624kjr\nHost: minio-c2.minio.io:31001\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.0.3\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20210424T171816Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20210424/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-copy-source;x-amz-copy-source-if-match;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=412, headers=Accept-Ranges: bytes\nContent-Length: 416\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nETag: \"71cff0a060f852067e443ad1e24ae26c\"\nLast-Modified: Sat, 24 Apr 2021 17:18:16 GMT\nServer: MinIO\nVary: Origin\nX-Amz-Request-Id: 1678DADA15762A4D\nX-Xss-Protection: 1; mode=block\nDate: Sat, 24 Apr 2021 17:18:16 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:775), io.minio.MinioClient.execute(MinioClient.java:563), io.minio.MinioClient.executePut(MinioClient.java:904), io.minio.MinioClient.copyObject(MinioClient.java:1232), FunctionalTest.testCopyObjectMatchETag(FunctionalTest.java:1850), FunctionalTest.copyObject(FunctionalTest.java:2016), FunctionalTest.runObjectTests(FunctionalTest.java:3757), FunctionalTest.runTests(FunctionalTest.java:3783), FunctionalTest.main(FunctionalTest.java:3927)]"
}
(8/15) Running minio-js tests ... done in 46 seconds
(9/15) Running minio-py tests ... done in 2 minutes and 36 seconds
(10/15) Running s3cmd tests ... FAILED in 5 seconds
{
  "name": "s3cmd",
  "duration": "2708",
  "function": "test_put_object_multipart",
  "status": "FAIL",
  "error": "WARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Too many failures. Giving up on '/mint/data/datafile-65-MB'\nERROR: \nUpload of '/mint/data/datafile-65-MB' part 1 failed. Use\n  /usr/local/bin/s3cmd abortmp s3://s3cmd-test-bucket-26560/s3cmd-test-object-29838 af50611d-f64a-4d51-ad67-87017f5e08ee\nto abort the upload, or\n  /usr/local/bin/s3cmd --upload-id af50611d-f64a-4d51-ad67-87017f5e08ee put ...\nto continue the upload.\nERROR: Upload of '/mint/data/datafile-65-MB' failed too many times (Last reason: )"
}
(10/15) Running s3select tests ... done in 5 seconds
(11/15) Running security tests ... done in 1 seconds

Executed 11 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

--- a/cmd/common-main.go
+++ b/cmd/common-main.go
@@ -341,28 +341,37 @@ func handleCommonEnvVars() {
 	case env.IsSet(config.EnvKMSMasterKey) && env.IsSet(config.EnvKESEndpoint):
 		logger.Fatal(errors.New("ambigious KMS configuration"), fmt.Sprintf("The environment contains %q as well as %q", config.EnvKMSMasterKey, config.EnvKESEndpoint))
 	}
-	if env.IsSet(config.EnvKMSSecretKey) {
-		KMS, err := kms.Parse(env.Get(config.EnvKMSSecretKey, ""))
-		if err != nil {
-			logger.Fatal(err, "Unable to parse the KMS secret key inherited from the shell environment")
-		}
-		GlobalKMS = KMS
-	} else if env.IsSet(config.EnvKMSMasterKey) {
-		logger.LogIf(GlobalContext, errors.New("legacy KMS configuration"), fmt.Sprintf("The environment variable %q is deprecated and will be removed in the future", config.EnvKMSMasterKey))
 
-		v := strings.SplitN(env.Get(config.EnvKMSMasterKey, ""), ":", 2)
-		if len(v) != 2 {
-			logger.Fatal(errors.New("invalid "+config.EnvKMSMasterKey), "Unable to parse the KMS secret key inherited from the shell environment")
-		}
-		secretKey, err := hex.DecodeString(v[1])
+	parseMasterKey := func(key string) error {
+		KMS, err := kms.Parse(env.Get(key, ""))
 		if err != nil {
-			logger.Fatal(err, "Unable to parse the KMS secret key inherited from the shell environment")
-		}
-		KMS, err := kms.New(v[0], secretKey)
-		if err != nil {
-			logger.Fatal(err, "Unable to parse the KMS secret key inherited from the shell environment")
+			v := strings.SplitN(env.Get(key, ""), ":", 2)
+			if len(v) != 2 {
+				return errors.New("invalid " + key)
+			}
+			secretKey, err := hex.DecodeString(v[1])
+			if err != nil {
+				return err
+			}
+			KMS, err = kms.New(v[0], secretKey)
+			if err != nil {
+				return err
+			}
 		}
 		GlobalKMS = KMS
+		return nil
+	}
+
+	if env.IsSet(config.EnvKMSSecretKey) {
+		if err = parseMasterKey(config.EnvKMSSecretKey); err != nil {
+			logger.Fatal(err, "Unable to parse the KMS secret key inherited from the shell environment")
+		}
+	} else if env.IsSet(config.EnvKMSMasterKey) {
+		logger.LogIf(GlobalContext, errors.New("legacy KMS configuration"),
+			fmt.Sprintf("The environment variable %q is deprecated and will be removed in the future", config.EnvKMSMasterKey))
+		if err = parseMasterKey(config.EnvKMSMasterKey); err != nil {
+			logger.Fatal(err, "Unable to parse the KMS secret key inherited from the shell environment")
+		}
 	}
 	if env.IsSet(config.EnvKESEndpoint) {
 		kesEndpoints, err := crypto.ParseKESEndpoints(env.Get(config.EnvKESEndpoint, ""))

--- a/docs/kms/IAM.md
+++ b/docs/kms/IAM.md
@@ -1,0 +1,61 @@
+# KMS IAM/Config Encryption
+
+MinIO will soon release a change that re-works the encryption of IAM and
+configuration data. Currently, MinIO encrypts IAM data (user/temp. credentials,
+policies and other configuration data) with the cluster root credentials before
+storing it on the backend disks. After release `RELEASE.2021-04-22T15-44-28Z`
+onwards, MinIO will use the KMS provided keys to encrypt the IAM data instead
+of the cluster root credentials. If the KMS is not enabled, MinIO will store
+the IAM data as plain text in its backend.
+
+### FAQ
+
+> Why is this change needed? Was the previous encryption not secure (enough)?
+
+The previous mechanism of encryption using the cluster root credentials itself
+is secure. The purpose of this change is to improve the encryption key
+management by delegating it to the KMS. Once the cluster root credentials are no
+longer needed to decrypt persistent data at the backend, they can be rotated
+much more easily. By using a KMS to protect the IAM data a potential security or
+compliance team has much more control over how/when a MinIO cluster can be
+started.
+
+> Does this mean I need an enterprise KMS setup to run MinIO (securely)?
+
+No, MinIO does not depend on any third-party KMS provider. You have three options here:
+
+- Run MinIO without a KMS. In this case all IAM data will be stored in plain-text.
+
+- Run MinIO with a single secret key. MinIO supports a static cryptographic key
+  that can act as minimal KMS. With this method all IAM data will be stored
+  encrypted. The encryption key has to be passed as environment variable.
+
+- Run MinIO with KES (minio/kes) in combination with any supported KMS as
+  secure key store. For example, you can run MinIO + KES + Hashicorp Vault.
+
+> What about an exiting MinIO deployment? Can I just upgrade my cluster?
+
+Yes, MinIO will try to transparently migrate any existing IAM data and either stores
+it in plaintext (no KMS) or re-encrypts using the KMS.
+
+> Is this change backward compatible? Will it break my setup?
+
+This change is not backward compatible for all setups. In particular, the native
+Hashicorp Vault integration - which has been deprecated already - won't be
+supported anymore. KES is now mandatory if a third-party KMS should be used.
+
+Further, since the configuration data is encrypted with the KMS, the KMS
+configuration itself can no longer be stored in the MinIO config file and
+instead must be provided via environment variables. If you have set your KMS
+configuration using e.g. the `mc admin config` commands you will need to adjust
+your deployment.
+
+Even though this change is backward compatible we do not expect that it affects
+the vast majority of deployments in any negative way.
+
+> Will an upgrade of an existing MinIO cluster impact the SLA of the cluster or will it even cause downtime?
+
+No, an upgrade should not cause any downtime. However, on the first startup -
+since MinIO will attempt to migrate any existing IAM data - the boot process may
+take slightly longer, but may not be visibly noticeable. Once the migration has
+completed, any subsequent restart should be as fast as before or even faster.

--- a/pkg/kms/single-key.go
+++ b/pkg/kms/single-key.go
@@ -28,7 +28,9 @@ import (
 	"fmt"
 	"strconv"
 	"strings"
+	"unicode/utf8"
 
+	"github.com/minio/sio"
 	"github.com/secure-io/sio-go/sioutil"
 	"golang.org/x/crypto/chacha20"
 	"golang.org/x/crypto/chacha20poly1305"
@@ -55,7 +57,7 @@ func Parse(s string) (KMS, error) {
 }
 
 // New returns a single-key KMS that derives new DEKs from the
-// given key. The given key must always be 32 bytes.
+// given key.
 func New(keyID string, key []byte) (KMS, error) {
 	if len(key) != 32 {
 		return nil, errors.New("kms: invalid key length " + strconv.Itoa(len(key)))
@@ -168,9 +170,37 @@ func (kms secretKey) GenerateKey(keyID string, context Context) (DEK, error) {
 	}, nil
 }
 
+func (kms secretKey) legacyDecryptKey(keyID string, sealedKey []byte, ctx Context) ([]byte, error) {
+	var derivedKey = kms.deriveKey(keyID, ctx)
+
+	var key [32]byte
+	out, err := sio.DecryptBuffer(key[:0], sealedKey, sio.Config{Key: derivedKey[:]})
+	if err != nil || len(out) != 32 {
+		return nil, err // TODO(aead): upgrade sio to use sio.Error
+	}
+	return key[:], nil
+}
+
+func (kms secretKey) deriveKey(keyID string, context Context) (key [32]byte) {
+	if context == nil {
+		context = Context{}
+	}
+	ctxBytes, _ := context.MarshalText()
+
+	mac := hmac.New(sha256.New, kms.key[:])
+	mac.Write([]byte(keyID))
+	mac.Write(ctxBytes)
+	mac.Sum(key[:0])
+	return key
+}
+
 func (kms secretKey) DecryptKey(keyID string, ciphertext []byte, context Context) ([]byte, error) {
 	if keyID != kms.keyID {
 		return nil, fmt.Errorf("kms: key %q does not exist", keyID)
+	}
+
+	if !utf8.Valid(ciphertext) {
+		return kms.legacyDecryptKey(keyID, ciphertext, context)
 	}
 
 	var encryptedKey encryptedKey

